plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java

```diff
@@ -19,74 +19,94 @@
-import com.google.api.client.util.ExponentialBackOff;
-import com.google.api.services.storage.Storage;
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.auth.oauth2.ServiceAccountCredentials;
+import com.google.cloud.http.HttpTransportOptions;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageOptions;
+
 import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.collect.MapBuilder;
 import org.elasticsearch.common.component.AbstractComponent;
 import org.elasticsearch.common.unit.TimeValue;
 import org.elasticsearch.env.Environment;
-
 import java.io.IOException;
+import java.net.URI;
+import java.security.GeneralSecurityException;
 import java.util.Map;
 
 public class GoogleCloudStorageService extends AbstractComponent {
 
     /** Clients settings identified by client name. */
     private final Map<String, GoogleCloudStorageClientSettings> clientsSettings;
 
-    public GoogleCloudStorageService(final Environment environment, final Map<String, GoogleCloudStorageClientSettings> clientsSettings) {
+    public GoogleCloudStorageService(Environment environment, Map<String, GoogleCloudStorageClientSettings> clientsSettings) {
         super(environment.settings());
         this.clientsSettings = clientsSettings;
     }
 
     /**
      * Creates a client that can be used to manage Google Cloud Storage objects.
      *
-     * @param clientName name of client settings to use from secure settings
+     * @param clientName
+     *            name of client settings to use from secure settings
      * @return a Client instance that can be used to manage Storage objects
      */
-    public Storage createClient(final String clientName) throws Exception {
+    public Storage createClient(final String clientName) throws GeneralSecurityException, IOException {
         final GoogleCloudStorageClientSettings clientSettings = clientsSettings.get(clientName);
         if (clientSettings == null) {
-            throw new IllegalArgumentException("Unknown client name [" + clientName + "]. Existing client configs: " +
-                Strings.collectionToDelimitedString(clientsSettings.keySet(), ","));
+            throw new IllegalArgumentException("Unknown client name [" + clientName + "]. Existing client configs: "
+                    + Strings.collectionToDelimitedString(clientsSettings.keySet(), ","));
         }
-
-        HttpTransport transport = GoogleNetHttpTransport.newTrustedTransport();
-        HttpRequestInitializer requestInitializer =  createRequestInitializer(clientSettings);
-
-        Storage.Builder storage = new Storage.Builder(transport, JacksonFactory.getDefaultInstance(), requestInitializer);
-        if (Strings.hasLength(clientSettings.getApplicationName())) {
-            storage.setApplicationName(clientSettings.getApplicationName());
+        final NetHttpTransport netHttpTransport = GoogleNetHttpTransport.newTrustedTransport();
```

Comment:
I have coded smth similar. I have not copied it verbatim because `URL#getFile` is a forbiddenApi and a URI conversion was required.
```
    private static HttpTransport createHttpTransport(final String endpoint) throws Exception {
        final NetHttpTransport.Builder builder = new NetHttpTransport.Builder();
        // requires java.lang.RuntimePermission "setFactory"
        builder.trustCertificates(GoogleUtils.getCertificateTrustStore());
        if (Strings.hasLength(endpoint)) {
            final URL endpointUrl = URI.create(endpoint).toURL();
            builder.setConnectionFactory(new DefaultConnectionFactory() {
                @Override
                public HttpURLConnection openConnection(final URL originalUrl) throws IOException {
                    // test if the URL is built correctly, ie following the `host` setting
                    if (originalUrl.getHost().equals(endpointUrl.getHost()) && originalUrl.getPort() == endpointUrl.getPort()
                            && originalUrl.getProtocol().equals(endpointUrl.getProtocol())) {
                        return super.openConnection(originalUrl);
                    }
                    // override connection URLs because some don't follow the config. See
                    // https://github.com/GoogleCloudPlatform/google-cloud-java/issues/3254 and
                    // https://github.com/GoogleCloudPlatform/google-cloud-java/issues/3255
                    URI originalUri;
                    try {
                        originalUri = originalUrl.toURI();
                    } catch (final URISyntaxException e) {
                        throw new RuntimeException(e);
                    }
                    String overridePath = "/";
                    if (originalUri.getRawPath() != null) {
                        overridePath = originalUri.getRawPath();
                    }
                    if (originalUri.getRawQuery() != null) {
                        overridePath += "?" + originalUri.getRawQuery();
                    }
                    return super.openConnection(
                            new URL(endpointUrl.getProtocol(), endpointUrl.getHost(), endpointUrl.getPort(), overridePath));
                }
            });
        }
        return builder.build();
    }
```